standard/variables.md

```diff
@@ -988,3 +998,185 @@ variable_reference
+
+> *Note*: The compiler determines the safe context through a static analysis of the program text. The safe context reflects the lifetime of a variable at runtime. *end note*
+
+There are three valid ref-safe-contexts:
+
+- ***declaration-block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. Each nested block represents a different context.  The ref-safe-context of a local variable is the declaration-block in which it is declared. A local variable declared in a method has a ref-safe-context of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block after the referent, or a nested block.
+- ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-context of the fields of a `struct` type is function-member. A variable with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.
+- ***caller-context***: Member fields of a `class` type and `ref` parameters have ref-safe-context of caller-context. A variable with ref-safe-context of caller-context can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.
+
+These values form a nesting relationship from narrowest (declaration-block) to widest (caller-context). Each nested block represents a different context.
+
+> *Example*: The following code shows examples of the different ref-safe-contexts. The declarations show the ref-safe-context for a referent to be the initializing expression for a `ref` variable. The examples show the ref-safe-context for a reference return:
+>
+> ```csharp
+> public class C
+> {
+>     // ref safe context of arr is "caller-context". 
+>     // ref safe context of arr[i] is "caller-context".
+>     private int[] arr = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }; 
+> 
+>     // ref safe context is "caller-context"
+>     public ref int M1(ref int r1)
+>     {
+>         return ref r1; // r1 is safe to ref return
+>     }
+>
+>     // ref safe context is "function-member"
+>     public ref int M2(int v1)
+>     {
+>         return ref v1; // error: v1 isn't safe to ref return
+>     }
+>
+>     public ref int M3()
+>     {
+>         int v2 = 5;
+> 
+>         return ref arr[v2]; // arr[v2] is safe to ref return
+>     }
+> 
+>     public void M4(int p) 
+>     {
+>         int v3 = 6;
+>
+>         // context of r2 is block, ref safe context of p is method
+>         ref int r2 = ref p;
+>
+>         // context of r3 is block, ref safe context of v3 is block
+>         ref int r3 = ref v3;
+>
+>         // context of r4 is block, ref safe context of arr[v3] is caller method
+>         ref int r4 = ref arr[v3]; 
+>     }
```

Comment:
>} [](http://example.com/codeflow?start=6&length=1)

Should the comments within `M4()` use the exact terms defined above: "declaration-block", "caller-context", and "function-member", rather than block, caller method, and method?